content/articles/web-apps.md

```diff
@@ -0,0 +1,108 @@
+
+- Tested: It's also good if the research code can have at least a basic suite of tests which run under CI - it's often difficult to test exact output values from research code (e.g. if it is stochastic), but it's useful to at least test that results are being successfully returned in the expected format.
+
+- Has a way of managing long-running processes: if research code invoked by the user from the application is going to take more than a couple of seconds to run, then it will need to run asynchronously of the initial web request. The web application will need some way to check status and progress while the code is running, and to notify the user when it is complete. We can work with the research team to design the best approach for the application.
+
+In addition, the research team will need to provide us with a good idea of the purpose, users and requirements of the application. It's ok for there to be multiple user groups, but we need to have a clear idea of who they are and how we will separate functionality for them. If there is a Shiny prototype we may use that as a starting point - if not, we will probably start by creating some mockups to agree on the general flow and functionality of the app.
+
+
+#  Application Patterns
+
+Almost all of our applications involve the user providing some input (data files, parameter values), triggering some analysis and then observing or downloading results. Different research and different user motivations calls for slightly different user experiences in the app. Here are some examples of different styles of application we have written:
+
+## Long-running model
+
+[HINT](https://naomi.unaids.org) runs the Naomi model for HIV indicators. The user uploads survey and other data, or imports it automatically from the AIDS Data Repository, and sets model parameters. The user then runs the model code, which can take several minutes. The web app polls the interface layer and provides status updates while the model is running. The user can then explore the output data in the web app, and can download output and summary data too.
+![HINT](/img/webapp-hint.png)
+
+## Pre-run model for intervention strategy
+
+[MINT](https://mint.dide.ic.ac.uk/) provides the user with access to a pre-run dataset of expected malaria prevalence and cases averted given various intervention options and regional characteristics. The app is tailored to guide decision making on best allocation of a budget. Because the data is pre-calculated, there are no long-running processes to manage.
+![MINT](/img/webapp-mint.png)
+
+## Model exploration
+
+The WODIN application allows the user to create, edit and explore a model written in [ODIN](https://mrc-ide.github.io/odin/). While this is primarily being used as a teaching tool, it could also be used to present a model (it can include default model code) and allow the user to run it, fit it to data, try their own code tweaks etc.
```

Comment:
```suggestion
[WODIN](https://epimodels.dide.ic.ac.uk/demo/) allows the user to create, edit and explore a model written in [ODIN](https://mrc-ide.github.io/odin/). While this is primarily being used as a teaching tool, it could also be used to present a model (it can include default model code) and allow the user to run it, fit it to data, try their own code tweaks etc.
```